Clean-up navigation

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

1. Navigation as a result of following a hyperlink should be done in a task: https://html.spec.whatwg.org/multipage/links.html#following-hyperlinks:dom-manipulation-task-source
2. The javascript url navigation should also be done in a task: https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents:dom-manipulation-task-source
3. In `window.load_url`, it seems there is no need to send a message to the script-thread(the entirety of `load_url` should instead be done in a task when appropriate), so we can just do that last part "sync" by calling a method on the script, which will send a message to the constellation(for the parallel navigation steps), or queue task(for the JS navigation), as appropriate. 
4. Separate the "normal" navigation flow from the handling of "navigate an iframe" message from constellation, since doing everything in one method as was previously done with `handle_navigate`, is confusing. 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23368)
<!-- Reviewable:end -->

gterzian committed:
clean-up navigation

security: check target and source origin before executing JS url

implement replacement-enabled flag as a HistoryEntryReplacement enum

add source origin string on loaddata

add LoadOrigin

iframe: remove optional load-data

auxiliaries: add load-data into info

constellation: remove url from Pipeline::new

check load origin: link to whatwg issue

switch loadorigin toplevel to constellation

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/constellation.rs, components/script/dom/htmlanchorelement.rs, components/script/dom/htmlformelement.rs, components/script/script_thread.rs, components/script/dom/window.rs
 * @cbrewster: components/constellation/constellation.rs
 * @paulrouget: components/constellation/constellation.rs
 * @kichjang: components/script/dom/htmlanchorelement.rs, components/script/dom/htmlformelement.rs, components/script/script_thread.rs, components/script/dom/window.rs, components/script_traits/lib.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

gterzian commented:
@bors-servo try-wpt

bors-servo commented:
:hourglass: Trying commit b8b9594ff8865ab29c01d6d30dd1d29100039d4f with merge 5f094b903c614a57a96b786e486c9bc0280e7489...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12211)

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 495170e2f8725d3512b09cc77607912a0f0a9d96 with merge 47d500eec681fd4fde9ff5c5faa7d741795ad110...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12212)

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 12bee2b90df4709963f90a9ecf94d1a471519b74 with merge 83dfb9ae113a8c22412ad319e34787631a967939...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12213)

gterzian commented:
Fixed a test by removing the task-queuing for the JS navigation I previously had introduced. It seems to be an issue of some contention: https://github.com/whatwg/html/issues/4427

bors-servo commented:
:hourglass: Trying commit 958c32562f042fee39869182f06c26ccebc83ea3 with merge 94846b7f379ac3425c16b2432a3acd4b590676e7...

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/16801.

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1557758701250.

servo-wpt-sync commented:
No upstreamable changes; closed existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16801.

Manishearth commented on components/script/dom/htmlanchorelement.rs:
```diff
@@ -667,7 +669,15 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
+        });
+        target_window
+            .task_manager()
+            .dom_manipulation_task_source()
```
should this be the networking task source?

gterzian replied:
The DOM manipulation is the right one: https://html.spec.whatwg.org/multipage/links.html#following-hyperlinks:dom-manipulation-task-source

Manishearth commented on components/script/dom/window.rs:
```diff
@@ -1722,17 +1722,18 @@ impl Window {
     }
 
     /// Commence a new URL load which will either replace this window or scroll to a fragment.
+    /// https://html.spec.whatwg.org/multipage/#navigating-across-documents
```
blank comment line between these two

Manishearth commented on components/script/dom/window.rs:
```diff
@@ -1790,13 +1791,20 @@ impl Window {
-                .unwrap();
+            // TODO: step 11, navigationType.
+            // Step 12, 13
+            let load_data = if let Some(mut load_data) = load_data {
```
Should we just replace the parameters of this function with a single `load_data`? Is there a reason to have both?

gterzian replied:
Yes that's a good idea...

Manishearth commented:
let me know when this is ready for a full review!

gterzian commented:
@bors-servo try=wpt (since I didn't get the result from last run, and it's gone from the grid)

bors-servo commented:
:hourglass: Trying commit 0bcdb62834d3c1eed3f9e481fd5dadbd67283a5f with merge b0ff0e4dae0cfb7cc54040acd68394c6674a30ea...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12228)

gterzian commented:
@Manishearth Ok, this one is now ready for a full review. 

The changes to the test should be seen in the light of https://github.com/whatwg/html/issues/3730#issuecomment-492071447

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/16812.

gterzian commented:
@bors-servo try=wpt (hoping something went from fail to pass)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

bors-servo commented:
:hourglass: Trying commit 4dbbd99f5618825886dc68a77f21e3778f74053f with merge 95f06ab5289ac5a7ec869b01faf6baad13f0d5b7...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12230)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
Looks like that `PASS` is intermittent...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@Manishearth Ok, this one is now ready for a full review.

The changes to the test should be seen in the light of [whatwg/html#3730 (comment)](https://github.com/whatwg/html/issues/3730#issuecomment-492071447)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@Manishearth Ok, this one is now ready for a full review.

The changes to the test should be seen in the light of [whatwg/html#3730 (comment)](https://github.com/whatwg/html/issues/3730#issuecomment-492071447)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@Manishearth Ok, this one is now ready for a full review.

The changes to the test should be seen in the light of [whatwg/html#3730 (comment)](https://github.com/whatwg/html/issues/3730#issuecomment-492071447)

I've also added a commit to at least make a beginning of security checks before executing a JS url...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@Manishearth Ok, this one is now ready for a full review.

The changes to the test should be seen in the light of [whatwg/html#3730 (comment)](https://github.com/whatwg/html/issues/3730#issuecomment-492071447)

I've also added a commit to at least make a beginning of security checks before executing a JS url...

gterzian commented:
The failure on travis is https://github.com/servo/servo/issues/23168

gterzian commented:
@Manishearth Ok, this one is now ready for a full review.

gterzian commented:
@Manishearth Ok, this one is now ready for a full review.

CYBAI commented:
@gterzian maybe you want to remove `[WIP]` in PR title 👀?

servo-wpt-sync commented:
PR title changed; changed existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@jdm I'm wondering if it's worth putting the changes to wpt test in a separate commit, since "clean-up navigation" doesn't mean that much in the context of wpt. What practice has been adopted for this?

jdm commented:
I am not aware of any guidance or common practice around upstreamed commits.

servo-wpt-sync commented:
PR title changed; changed existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
Still "ready for review" by the way, just removed that part from the title since there is already a label for that... @Manishearth

Manishearth commented:
Yeah, I plan to get to this, I'm just not super familiar with the spec
here. Apologies for the delay.

On Mon, May 27, 2019, 3:31 AM Gregory Terzian <notifications@github.com>
wrote:

> Still "ready for review" by the way, just removed that part from the title
> since there is already a label for that... @Manishearth
> <https://github.com/Manishearth>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/servo/servo/pull/23368?email_source=notifications&email_token=AAMK6SEIGN2LXPH5URV4K6LPXOZ7XA5CNFSM4HMKCYAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWJOJRA#issuecomment-496166084>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMK6SEHM6NHNDR3DL2W3J3PXOZ7XANCNFSM4HMKCYAA>
> .
>

Manishearth reviewed (approved):
This seems good.

@asajeffrey can you also have a look?

asajeffrey commented on components/script/script_thread.rs:
```diff
@@ -869,6 +867,52 @@ impl ScriptThread {
+            };
+            let script_thread = unsafe { &*script_thread };
+            let is_javascript = load_data.url.scheme() == "javascript";
+            // If resource is a request whose url's scheme is "javascript"
```
Spec link? to https://html.spec.whatwg.org/multipage/browsing-the-web.html#javascript-protocol

asajeffrey commented on components/script/script_thread.rs:
```diff
@@ -869,6 +867,52 @@ impl ScriptThread {
 
+    /// Step 13 of https://html.spec.whatwg.org/multipage/#navigate
+    pub fn navigate(parent_pipeline_id: PipelineId, mut load_data: LoadData, replace: bool) {
+        SCRIPT_THREAD_ROOT.with(|root| {
```
Rename `parent_pipeline_id` (which is normally used to mean the pipeline id of an iframe's parent) to `pipeline_id`.

asajeffrey commented on tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/execution-timing/029.html:
```diff
@@ -25,13 +28,24 @@
                 assert_any(assert_array_equals, eventOrder, [
-                  ['inline script #1', 'end script #1', 'JS URL', 'inline script #2'],
-                  ['inline script #1', 'end script #1', 'inline script #2', 'JS URL']]);
+                  ['inline script #1', 'end script #1', 'beforeunload event', 'inline script #2', 'JS URL'],
```
The order of the `JS URL` and `inline script #2` events have changed?

gterzian replied:
The test, before these changes, assumed that `JS URL` happens in the same task as the one in which we initiate the navigation, while the spec says to queue a subsequent task for the `JS URL` part. So that part is now guaranteed to be last, and I've added the `beforeunload event` part since that will happen in the "first" task where the navigation "starts". 

So basically 'beforeunload event' has taken the place of `JS URL`, while 'inline script #2' could fire either before or after it, and `JS URL` is now guaranteed to require a subsequent task. 

See the discussion at https://github.com/whatwg/html/issues/3730#issuecomment-492071447

asajeffrey replied:
Might be worth putting a link to that issue in the test?

gterzian replied:
Ok, done.

asajeffrey reviewed (commented):
Mostly LGTM, some questions though.

gterzian commented on components/script/dom/location.rs:
```diff
@@ -38,15 +40,24 @@ impl Location {
+        let load_data = LoadData::new(url, Some(pipeline_id), Some(referrer), referrer_policy);
+        // TODO: rethrow exceptions, set exceptions enabled flag.
+        self.window.load_url(false, false, load_data);
+    }
```
Oh yeah, I actually missed how those booleans differed per call, good point with regards to using enums instead...

gterzian replied:
I fixed the initial commit, and then added a comment to replace one flag with an enum.

asajeffrey replied:
Thanks.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented on components/script/dom/htmlformelement.rs:
```diff
@@ -530,27 +531,43 @@ impl HTMLFormElement {
+                .root()
+                .load_url(
+                    false,
+                    false,
```
Ok, for one flag I added a commit to introduce a `HistoryEntryReplacement` enum, and for the other flag I filed https://github.com/servo/servo/issues/23481

asajeffrey replied:
Ta.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
The failure on CI appears unrelated https://travis-ci.com/servo/servo/jobs/204407647#L2788

```
Caused by:
  failed to get 200 response from `https://static.crates.io/crates/binjs_meta/binjs_meta-0.4.3.crate`, got 502
make[4]: *** [force-cargo-host-program-build] Error 101
make[3]: *** [js/src/frontend/binsource/host] Error 2
make[2]: *** [compile] Error 2
make[1]: *** [default] Error 2
make: *** [all] Error 2
thread 'main' panicked at 'assertion failed: result.success()', /home/travis/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/e21c05b/build.rs:117:5
```

gterzian commented:
@asajeffrey should we try to merge it?

asajeffrey commented:
Sure, squash and then r=me

gterzian commented:
Ok squashed

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

jdm commented:
@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit d2a254e has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey d2a254e2079bbb33872125b9868318a6095478a3 -->

bors-servo commented:
:hourglass: Testing commit d2a254e2079bbb33872125b9868318a6095478a3 with merge 9e53058acbbff565eb5f1c29816106662da91261...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12460)

CYBAI commented:
<details>

```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "iframe.contentDocument is null", 
    "stack": "@http://web-platform.test:8000/dom/nodes/Document-contentType/contentType/contenttype_javascripturi.html:10:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\nTest.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1634:17\n", 
    "subtest": "Javascript URI document.contentType === 'text/html'", 
    "test": "/dom/nodes/Document-contentType/contentType/contenttype_javascripturi.html", 
    "line": 79251, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-2.html?1234\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-3.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-3.html?pipe=sub:14:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-3.html?pipe=sub:13:1\n", 
    "subtest": "HTTP Referer should use the pushed state", 
    "test": "/html/browsers/history/the-history-interface/009.html", 
    "line": 87886, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-2.html?1234\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-3.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-3.html?pipe=sub:17:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-3.html?pipe=sub:16:1\n", 
    "subtest": "document.referrer should use the pushed state", 
    "test": "/html/browsers/history/the-history-interface/009.html", 
    "line": 87887, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-4.html?2345\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-5.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-5.html?pipe=sub:14:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-5.html?pipe=sub:13:1\n", 
    "subtest": "HTTP Referer should use the replaced state", 
    "test": "/html/browsers/history/the-history-interface/009.html", 
    "line": 87888, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-4.html?2345\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/009-5.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-5.html?pipe=sub:17:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/009-5.html?pipe=sub:16:1\n", 
    "subtest": "document.referrer should use the replaced state", 
    "test": "/html/browsers/history/the-history-interface/009.html", 
    "line": 87889, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_array_equals: lengths differ, expected 5 got 4", 
    "stack": "onload</<@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/079.html:18:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1618:20\n", 
    "subtest": " setting location to javascript URL from event handler ", 
    "test": "/html/semantics/scripting-1/the-script-element/execution-timing/079.html", 
    "line": 89166, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/source/support/location-set.html\" but got \"http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/source/support/support/dummy.html\"", 
    "stack": "onload</fr.onload<@http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html:15:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\nTest.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1634:17\n", 
    "subtest": "Set location from a function called from a parent", 
    "test": "/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html", 
    "line": 96297, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-2.html?1234\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-3.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-3.html?pipe=sub:14:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-3.html?pipe=sub:13:1\n", 
    "subtest": "HTTP Referer should use the pushed state (before onload)", 
    "test": "/html/browsers/history/the-history-interface/010.html", 
    "line": 98712, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-2.html?1234\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-3.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-3.html?pipe=sub:17:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-3.html?pipe=sub:16:1\n", 
    "subtest": "document.referrer should use the pushed state (before onload)", 
    "test": "/html/browsers/history/the-history-interface/010.html", 
    "line": 98713, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-4.html?2345\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-5.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-5.html?pipe=sub:14:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-5.html?pipe=sub:13:1\n", 
    "subtest": "HTTP Referer should use the replaced state (before onload)", 
    "test": "/html/browsers/history/the-history-interface/010.html", 
    "line": 98714, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-4.html?2345\" but got \"http://web-platform.test:8000/html/browsers/history/the-history-interface/010-5.html?pipe=sub\"", 
    "stack": "@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-5.html?pipe=sub:17:9\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/html/browsers/history/the-history-interface/010-5.html?pipe=sub:16:1\n", 
    "subtest": "document.referrer should use the replaced state (before onload)", 
    "test": "/html/browsers/history/the-history-interface/010.html", 
    "line": 98715, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "noopener window.open targeting _top", 
    "test": "/html/browsers/the-window-object/window-open-noopener.html?_top", 
    "line": 119973, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/html/browsers/the-window-object/window-open-noopener.html?_top", 
    "line": 119974, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "noopener window.open targeting _self", 
    "test": "/html/browsers/the-window-object/window-open-noopener.html?_self", 
    "line": 122854, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/html/browsers/the-window-object/window-open-noopener.html?_self", 
    "line": 122855, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Setting location from document.write() call should not trigger load event until that load completes", 
    "test": "/html/webappapis/dynamic-markup-insertion/closing-the-input-stream/load-event-after-location-set-during-write.window.html", 
    "line": 131644, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "noopener window.open targeting _parent", 
    "test": "/html/browsers/the-window-object/window-open-noopener.html?_parent", 
    "line": 143864, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/html/browsers/the-window-object/window-open-noopener.html?_parent", 
    "line": 143865, 
    "action": "test_result", 
    "expected": "OK"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_array_equals: lengths differ, expected 4 got 3\n\nassert_array_equals: lengths differ, expected 4 got 3", 
    "stack": "test@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/028.html:21:17\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1618:20\n", 
    "subtest": " scheduler: javascript: URL", 
    "test": "/html/semantics/scripting-1/the-script-element/execution-timing/028.html", 
    "line": 146338, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_array_equals: lengths differ, expected 5 got 4\n\nassert_array_equals: lengths differ, expected 5 got 4", 
    "stack": "onload/<@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/030.html:31:13\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1618:20\n", 
    "subtest": " scheduler: javascript: URL in HREF, onclick handler", 
    "test": "/html/semantics/scripting-1/the-script-element/execution-timing/030.html", 
    "line": 146960, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_array_equals: lengths differ, expected 5 got 4\n\nassert_array_equals: lengths differ, expected 5 got 4", 
    "stack": "final_test@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/execution-timing/029.html:35:17\nTest.prototype.step_timeout/<@http://web-platform.test:8000/resources/testharness.js:1652:20\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1594:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1618:20\n", 
    "subtest": " scheduler: javascript: URL in HREF", 
    "test": "/html/semantics/scripting-1/the-script-element/execution-timing/029.html", 
    "line": 197319, 
    "action": "test_result", 
    "expected": "PASS"
}
```

</details>

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@bors-servo try=wpt

So the problem as the origin check before executing the JS url. It turns out `ServoUrl` contains an `Arc`, so by the time the check was made, that url was turned into `Opaque` and thus it always failed to match the origin of the target document. 

I've added a field to `LoadData` containing the string representing the origin at the time that the navigation was initiated, so it can be used reliably for the comparison.

bors-servo commented:
:hourglass: Trying commit 2dcad34d5aae83aa2fa8f9358b56289328bd3197 with merge e4308f9670adf4d92dab23c22fcd944543c813b7...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12463)

gterzian commented:
@bors-servo try=wpt (forgot to update the iframe origin check)

bors-servo commented:
:hourglass: Trying commit 9fc37e6f0e977486aea87b6381928d2043aee66a with merge da2d9de76cdd85b2c51ddcc4ad8ed06dd81bebec...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

asajeffrey commented:
@gterzian Can we use ImmutableOrigin rather than a string?

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12466)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@asajeffrey Could you please take a look at the last few commits? There were a few issues introduced by the origin check for the JS url, and by the refactoring of the navigation code in `location`. 

I also ended-up adding one separate commit to perform some checks when setting `location.protocol`, initially because it seemed like my changes made a test fail which previously passed, but actually that was just because I was on an older master and the test was marked as failing on the latest too. In relation I've also filed https://github.com/servo/servo/issues/23522 since my current solution is probably not optimal.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 76ba7bb557eef988be01663e05bd0583a4d199e7 with merge 09859a05356396f5b44dd602e82605a82e83ad00...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12472)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@bors-servo try-wpt (that "extra" commit turned into quit a few problems, now hopefully fixed, and there was still an issue with the opaque origins of about:blank pages always differing, which made the same-origin check for the JS url fail with `window.open`)

bors-servo commented:
:hourglass: Trying commit b5afe40dfede3d02107dd4ce22502bfa2ccfc288 with merge d8695b79b4cd4180a14ddf6b6c3bd1d2fdcf1a87...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12531)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@bors-servo try=wpt (allowing JS url evaluation when the target has an opaque origing, which covers "about:blank" document such as an auxialiary or an iframe with no src).

bors-servo commented:
:hourglass: Trying commit 30d35ec31824e7020d0abcdc2e604bbf2b828f6a with merge fa137aa00a2af34235f9a2f86cb4e2541bee175e...

jdm commented:
That sounds very similar to the problem I described in https://github.com/servo/servo/issues/22879. Our handling of origins of about:blank right now is very incorrect. We should be separating the concept of a document origin from the origin of a URL.

jdm commented:
To be clear, I don't necessarily mean that work should be a prerequisite to merging these changes. It's one of the things that @avadacatavra is intending to look into.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12476)

asajeffrey commented on components/script/dom/htmlanchorelement.rs:
```diff
@@ -667,7 +682,24 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
-        debug!("following hyperlink to {}", url);
-        target_window.load_url(url, replace, false, referrer, referrer_policy);
+        let pipeline_id = target_window.upcast::<GlobalScope>().pipeline_id();
+        println!("Url: {:?}", url);
```
`println!` should be `debug!`

gterzian replied:
Yes I'm going to remove those print and console.logs...

asajeffrey commented on components/script_traits/lib.rs:
```diff
@@ -121,6 +122,9 @@ pub enum LayoutControlMsg {
 pub struct LoadData {
+    /// The origin of the page when it started the load.
+    /// Used for same-origin comparisons.
+    pub source_origin: ImmutableOrigin,
```
I'm confused, the comment makes it sound like this is the origin of the page that triggered a navigation, but when `LoadData::new` is called by the constellation, it's set to the origin of the URL.

gterzian replied:
Yes currently, it's a bit of a hack, and it's only meant for when the `LoadData` is created from script. The ones created in the constellation are I think either top-level loads initiated by the embedder, or part of the "create a new iframe or auxiliary" flows which I don't think involve any potential JS URL(those would happen in a later once the about:blank iframe or auxiliary has been created and is then navigated).

I think this needs revisiting as @jdm suggested, maybe with a new `NavigationOrigin` concept, that would have things like `TopLevel` and `Script(ImmutableOrigin or String of Origin)` and so on.

gterzian replied:
And by the way the goal is to do those JS url evaluations in a subsequent task queued from the navigation steps, and prevent the security problems noted at https://github.com/whatwg/html/issues/2591

gterzian replied:
See https://github.com/servo/servo/issues/22879

asajeffrey replied:
OK, this is possibly just bikeshedding the comment and field name.

asajeffrey commented on tests/wpt/metadata/html/browsers/history/the-location-interface/location-protocol-setter-non-broken.html.ini:
```diff
@@ -17,19 +17,3 @@
-
-  [Set HTTP URL frame location.protocol to x]
-    expected: FAIL
-
```
Yay!

asajeffrey commented on tests/wpt/web-platform-tests/html/browsers/history/the-location-interface/resources/post-your-protocol.html:
```diff
@@ -1,4 +1,5 @@
 <script>
   var id = location.search.substring(1);
   parent.postMessage({ id: id, protocol: location.protocol }, "*");
+  console.log("posting")
```
`console.log` in a WPT?

asajeffrey commented on tests/wpt/web-platform-tests/html/semantics/scripting-1/the-script-element/execution-timing/029.html:
```diff
@@ -10,6 +10,9 @@
         <p><a href="javascript:log('JS URL')"></a></p>
         <script>log('inline script #1');
+                window.addEventListener("beforeunload", function( event ) {
+                        log('beforeunload event');
```
Ditto.

jdm replied:
This is different from console.log.

asajeffrey replied:
Oops, didn't look very carefully at that!

asajeffrey reviewed (commented):
This is looking good! Really the only thing that's left is my confusion about the semantics of `source_origin` in `LoadData`.

gterzian commented:
The test failure is https://github.com/servo/servo/issues/23379

I'll clean up those console logs and printlns...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

asajeffrey commented:
I think this is now at a state where there's a bit of tidying up and squashing, but then it can land.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23154) made this pull request unmergeable. Please resolve the merge conflicts.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@asajeffrey could you please take a look at the last commit, I'm trying to convey more meaning to the load origin concept...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@asajeffrey Ok I squashed and I think I cleaned everything up. The only real change I made relates to adding this concept: https://github.com/servo/servo/pull/23368/files#diff-bf561a46a499a0c2dc837bea89df1be0R124

asajeffrey commented:
Not sure that `TopLevel` is the right name, e.g. for loads of iframes?

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
> Not sure that TopLevel is the right name, e.g. for loads of iframes?

Good point, please see the last commit(to be squashed).

gterzian commented:
Actually, looking at that last commit, I'm wondering if `TopLevel` was not better, since those are the "about:blank" loads for initial iframes and auxiliary, which then later can be navigated by script. So perhaps it's right to say that those initial blank pages are created "out of nowhere" by the constellation, and only later maybe navigated by script?

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@asajeffrey Could you please look at the last commit, I think I have found the solution to our problem stated above.

asajeffrey commented:
Sorry, traveling at the moment, hopefully I'll get a chance to look at this soon!

asajeffrey commented on components/script_traits/lib.rs:
```diff
@@ -117,10 +118,25 @@ pub enum LayoutControlMsg {
 #[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct LoadData {
+    /// The origin of the page when it started the load.
+    /// Used for same-origin comparisons.
```
The comment here disagrees with the comment in the definition of `LoadOrigin`. Is this the origin of the page being loaded, or the origin where the load was initiated?

gterzian replied:
Ok, I've update the comment, and it's supposed to be the origin where the load was initiated.

asajeffrey commented on components/script_traits/lib.rs:
```diff
@@ -117,10 +118,25 @@ pub enum LayoutControlMsg {
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum LoadOrigin {
+    /// A load originating in the constellation.
+    TopLevel,
```
This seems an odd name, since the constellation may be loading an iframe. Is the intension that this is for loads initiated by navigation and history traversal?

gterzian replied:
> Is the intension that this is for loads initiated by navigation and history traversal?
No that is not the intention.

The changes in this PR would only use this variant in the following three cases:

1. When the compositor tells us to load a new url for a given existing top-level(I guess corresponding to a user manually typing a url in the address bar and hitting return).
2. When the compositor tells us to load a new top-level for a given url(I guess corresponding to a user manually opening a new tab). 
3. When a top-level panics and we need to show an error page.

asajeffrey replied:
Might be worth naming those cases, e.g. `Navigation`, `Traversal`, `Initial`, `Error`? The current name `TopLevel` isn't ideal, since traversing the session history can end up loading a document inside an iframe.

gterzian replied:
Ok so I had missed that a `LoadData` was stored on the pipeline, and potentially used to reload the page as part of traversal if it had been evicted from the bfcache. 

And in that case, would a `LoadOriging::Traversal` be appropriate? It would require us to mutate the load-data I think, since the original load isn't going to be `Traversal`, right? Also, it might make sense to use the original loadata, since we're merely reloading the page according to the original parameters, which in some way should be transparent to script(which shouldn't be aware of whether a reload was necessary or not, since it should just get the same page originally stored in the bfcache?)?

So perhaps we should add variants for the various specific cases currently all grouped under `TopLevel`, and I'm wondering if reloading because of traversal should not keep the original `LoadOrigin::Script`. 

> since traversing the session history can end up loading a document inside an iframe

Is that a special case different from the general "reloading" of a page evicted from the bfcache?

gterzian replied:
How about `LoadOrigin::Constellation` for those cases where we actually call `LoadData::new` in the constellation, and `LoadOrigin::Script(origin)` for where we call it from script? Then a potential reload as part of history traversal will re-use the `LoadData` initially created from script, and only reloaded by the constellation, contain the original `LoadOrigin::Script(origin)`?

Also note that currently the concept is only used at https://github.com/servo/servo/pull/23368/files#diff-1879ac6bd5d567e2aa43529d33474677R862

asajeffrey commented on components/constellation/constellation.rs:
```diff
@@ -2300,7 +2312,13 @@ where
         // TODO: Referrer?
-        let load_data = LoadData::new(url.clone(), Some(parent_pipeline_id), None, None);
+        let load_data = LoadData::new(
+            LoadOrigin::Script(url.origin()),
```
Why is this the origin associated with `url` (which is always going to be a fresh opaque origin, since `url` is `about:blank`)?

gterzian replied:
Thanks, I switched to using the `LoadOrigin` that is sent as part of the `IFrameLoadInfo`.

asajeffrey commented on components/script/dom/htmlanchorelement.rs:
```diff
@@ -667,7 +682,23 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
-        target_window.load_url(url, replace, false, referrer, referrer_policy);
+        let pipeline_id = target_window.upcast::<GlobalScope>().pipeline_id();
+        let load_data = LoadData::new(
+            LoadOrigin::Script(document.url().origin()),
```
Should this be the origin associated with the document's URL, or the document's domain?

gterzian replied:
I think the document's domain is better, I switched to using that, thanks!

asajeffrey commented on components/script/dom/htmliframeelement.rs:
```diff
@@ -168,6 +175,7 @@ impl HTMLIFrameElement {
 
         let global_scope = window.upcast::<GlobalScope>();
         let load_info = IFrameLoadInfo {
+            load_origin: LoadOrigin::Script(document.url().origin()),
```
Ditto, should we use the URL's origin or `document.domain`?

asajeffrey commented on components/script/dom/htmliframeelement.rs:
```diff
@@ -270,6 +278,7 @@ impl HTMLIFrameElement {
 
         let document = document_from_node(self);
         let load_data = LoadData::new(
+            LoadOrigin::Script(document.url().origin()),
```
Ditto.

asajeffrey commented on components/script/dom/htmliframeelement.rs:
```diff
@@ -296,6 +309,7 @@ impl HTMLIFrameElement {
         let window = window_from_node(self);
         let pipeline_id = Some(window.upcast::<GlobalScope>().pipeline_id());
         let load_data = LoadData::new(
+            LoadOrigin::Script(document.url().origin()),
```
Ditto.

asajeffrey commented on components/script/dom/location.rs:
```diff
@@ -38,6 +40,28 @@ impl Location {
+        let referrer_policy = self.window.Document().get_referrer_policy();
+        let pipeline_id = self.window.upcast::<GlobalScope>().pipeline_id();
+        let load_data = LoadData::new(
+            LoadOrigin::Script(self.window.get_url().origin()),
```
Ditto.

asajeffrey commented on components/script/dom/location.rs:
```diff
@@ -217,7 +241,20 @@ impl LocationMethods for Location {
+        // TODO: avoid double parse,
+        // restructure set_url_component to accomodate various checks.
+        let mut try_parse = ServoUrl::from_url(url);
+        UrlHelper::SetProtocol(&mut try_parse, value.clone());
```
This looks like a different algorithm from `https://html.spec.whatwg.org/multipage/#dom-location-protocol`, does it achieve the same result?

gterzian replied:
I think so, yes. The call to `self.set_url_component` would correspond to Step 7, and these changes attempt to introduce the checks made at 4 to 6, before going to step 7. 

The problem is the double call to `UrlHelper::SetProtocol`, and the need to copy the url to do the check since we don't want to mutate it as part of it. 

There is an issue addressing this more comprehensively at https://github.com/servo/servo/issues/23522

asajeffrey replied:
OK, cite the issue and that's fine.

asajeffrey commented on components/script/script_thread.rs:
```diff
@@ -855,6 +855,72 @@ impl ScriptThread {
 
+    /// Check that two origins are "similar enough",
+    /// for now only used to prevent cross-origin JS url evaluation.
+    pub fn check_load_origin(source: &LoadOrigin, target: &ImmutableOrigin) -> bool {
```
It's not obvious to me how this translates into concepts from the spec.

gterzian replied:
Yes, that concept isn't in the spec yet, and this is more of an attempt to address the security issues in the current spec noted at https://github.com/whatwg/html/issues/2591

asajeffrey replied:
Cite that issue?

asajeffrey reviewed (commented):
I'm still confused about what `LoadOrigin` is, and how it relates to the spec.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@asajeffrey I think there is one last bit we could try to clean-up:

We have some code [here](https://github.com/servo/servo/blob/49bf3ba05f264a086fa5f421d5d40c6402ea5ada/components/script/script_thread.rs#L1237) and [here](https://github.com/servo/servo/blob/7490dd6f0deafc8192126a88a8a6f75c100e5d71/components/constellation/constellation.rs#L919) that checks if the load is `about:blank`, and if not, grabs the origin of the "creator" with various if/else. I think that "creator" origin could be simplified to be the `load_origin` of `LoadData`. Should we give this a try?

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@asajeffrey I've tried to address your comments, and also added a few commits for smaller refactorings which appeared reasonably around the `LoadData` concept, please take a look. I think they probably all could be squashed into "clean-up navigation".

asajeffrey commented:
Left a few comments. We're getting there!

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

gterzian commented:
@asajeffrey Ok so re our discussion on `LoadOrigin::TopLevel`, I tried switching that name to `LoadOrigin::Constellation`(see the last commit), and it seems to make sense for the following reasons:

1. It's only used when we actually do a `LoadData::New` inside the constellation, which is only in either embedder driven loads, or when a top-level panics and the constellation has to show an error page. 
2. That way `LoadOrigin::Script(origin)` is reserved for loads originating from script. That includes reloading a page evicted from the bf-cache, and my understanding is that the constellation in that case re-uses the stored original `LoadData`, whose origin will then be ``LoadOrigin::Script(origin)`, which I think makes more sense than for example switching that to being a dedicated `LoadOrigin::Traversal`. 

I've also removed my commit that related to the location protocol setter, since that is now being dealt with at https://github.com/servo/servo/pull/23670/

The current commits still require a squash.

Please let me know what you think.

jdm commented:
@asajeffrey Do you have time to finish up the review following the previous discussion?

asajeffrey commented:
Oh oops, this totally fell between the cracks. I'll have a look tomorrow, ping me if I forget again. (looks sheepish)

asajeffrey commented:
OK, this looks good to go now! Squash and r=me.

gterzian commented:
Ok, it's squashed. I cannot do the r thing.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#16812.

jdm commented:
@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit 571beec has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 571beec179fe9fd5fff2c12b3c5dfa0a5d93df01 -->

bors-servo commented:
:hourglass: Testing commit 571beec179fe9fd5fff2c12b3c5dfa0a5d93df01 with merge 334157b6e0e3bbaf07a00a2e4074cb92e5fb0aa1...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12840)

CYBAI commented:
```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Setting location from document.write() call should not trigger load event until that load completes", 
    "test": "/html/webappapis/dynamic-markup-insertion/closing-the-input-stream/load-event-after-location-set-during-write.window.html", 
    "line": 213332, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

gterzian commented:
@bors-servo retry (I think it's intermittent https://github.com/servo/servo/issues/23379)

gterzian commented:
@bors-servo retry

CYBAI commented:
@gterzian looks like @bors-servo is busy now 👀 maybe we need to wait for it

bors-servo commented:
:hourglass: Testing commit 571beec179fe9fd5fff2c12b3c5dfa0a5d93df01 with merge 1c1a4b164cc6cd2f683ad85043a8deeb8ddffa09...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12843)

CYBAI commented:
@bors-servo retry

- maybe new intermittent?

```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "first argument: absolute url", 
    "test": "/html/browsers/the-window-object/apis-for-creating-and-navigating-browsing-contexts-by-name/creating_browsing_context_test_01.html", 
    "line": 147652, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

bors-servo commented:
:hourglass: Testing commit 571beec179fe9fd5fff2c12b3c5dfa0a5d93df01 with merge 5fb6b4c6736926729f944e2f013781fe87ba9bec...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/WZ_pZsVxRECVCbQIbW53Cg)

jdm commented:
@bors-servo retry
* #23290 #23379

bors-servo commented:
:hourglass: Testing commit 571beec179fe9fd5fff2c12b3c5dfa0a5d93df01 with merge 9dca60f31f6a20b066bb475d142c18c453439207...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/GLZtdAfNRp-Kil4_Wz5lBA)

jdm commented:
@bors-servo retry
* #23290

bors-servo commented:
:bomb: Failed to start rebuilding: `Unknown error`

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12851), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12911), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/COZokrqYTeGHBaCZJqIAbg)
Approved by: asajeffrey
Pushing 2dfbbd0ca24ae17cb282a0f02289b42e9334947b to master...
